File: apps/backend/src/EnterpriseAgentOs.Domain/Models/PlanLimits.cs
namespace EnterpriseAgentOs.Domain.Models;

/// <summary>
/// Single source of truth for plan limits. Used by both billing enforcement
/// and the API response so the frontend never hardcodes limits.
/// </summary>
public static class PlanLimits
{
    // ── Individual plans ─────────────────────────────────────────────────────
    public static readonly PlanLimit IndividualFree = new("free",  1,    500_000L);
    public static readonly PlanLimit IndividualPro  = new("pro",   3, 10_000_000L);

    // ── Org plans ─────────────────────────────────────────────────────────────
    public static readonly PlanLimit OrgFree  = new("free",  1,    500_000L);
    public static readonly PlanLimit OrgTeam  = new("team", 10, 25_000_000L);
    // Enterprise limits are custom — stored on OrgSubscription directly.

    public static PlanLimit ForIndividualPlan(string plan) => plan switch
    {
        "pro" => IndividualPro,
        _     => IndividualFree,   // default to free
    };

    public static PlanLimit ForOrgPlan(string plan) => plan switch
    {
        "team"       => OrgTeam,
        "enterprise" => throw new ArgumentException("Enterprise limits are stored on OrgSubscription", nameof(plan)),
        _            => OrgFree,
    };
}

public sealed record PlanLimit(string Plan, int ConcurrentAgents, long CreditsPerMonth)
{
    /// <summary>Human-readable plan description derived from the actual limits.</summary>
    public string Description
    {
        get
        {
            var agentWord = ConcurrentAgents == 1 ? "agent" : "agents";
            var credits = CreditsPerMonth >= 1_000_000
                ? $"{CreditsPerMonth / 1_000_000M:#.##}M"
                : $"{CreditsPerMonth / 1_000M:#.##}k";
            return $"{ConcurrentAgents} concurrent {agentWord}, {credits} credits/month";
        }
    }
}

File: apps/backend-java/src/main/java/com/enterpriseagentos/backendjava/domain/features/management/registries/PlanLimits.java
package com.enterpriseagentos.backendjava.domain.features.management.registries;

import com.enterpriseagentos.backendjava.domain.common.*;
import com.enterpriseagentos.backendjava.domain.common.models.*;
import com.enterpriseagentos.backendjava.domain.common.primitives.*;
import com.enterpriseagentos.backendjava.domain.common.services.*;
import com.enterpriseagentos.backendjava.domain.common.valueobjects.*;
import com.enterpriseagentos.backendjava.domain.events.*;
import com.enterpriseagentos.backendjava.domain.features.agents.dtos.*;
import com.enterpriseagentos.backendjava.domain.features.agents.enums.*;
import com.enterpriseagentos.backendjava.domain.features.agents.interfaces.*;
import com.enterpriseagentos.backendjava.domain.features.agents.models.*;
import com.enterpriseagentos.backendjava.domain.features.agents.registries.*;
import com.enterpriseagentos.backendjava.domain.features.analytics.dtos.*;
import com.enterpriseagentos.backendjava.domain.features.analytics.enums.*;
import com.enterpriseagentos.backendjava.domain.features.analytics.interfaces.*;
import com.enterpriseagentos.backendjava.domain.features.analytics.mappers.*;
import com.enterpriseagentos.backendjava.domain.features.analytics.models.*;
import com.enterpriseagentos.backendjava.domain.features.management.dtos.*;
import com.enterpriseagentos.backendjava.domain.features.management.exceptions.*;
import com.enterpriseagentos.backendjava.domain.features.management.interfaces.*;
import com.enterpriseagentos.backendjava.domain.features.management.models.*;
import com.enterpriseagentos.backendjava.domain.features.management.registries.*;
import com.enterpriseagentos.backendjava.domain.features.management.services.*;
import com.enterpriseagentos.backendjava.domain.features.mcp.dtos.*;
import com.enterpriseagentos.backendjava.domain.features.mcp.enums.*;
import com.enterpriseagentos.backendjava.domain.features.mcp.interfaces.*;
import com.enterpriseagentos.backendjava.domain.features.mcp.models.*;

public final class PlanLimits {
    public static final PlanLimit INDIVIDUAL_FREE = new PlanLimit(SubscriptionPlan.Free, 1, 500_000L);
    public static final PlanLimit INDIVIDUAL_PRO = new PlanLimit(SubscriptionPlan.Pro, 3, 10_000_000L);
    public static final PlanLimit ORG_FREE = new PlanLimit(SubscriptionPlan.Free, 1, 500_000L);
    public static final PlanLimit ORG_TEAM = new PlanLimit(SubscriptionPlan.Team, 10, 25_000_000L);

    private PlanLimits() {
    }

    public static PlanLimit forIndividualPlan(SubscriptionPlan plan) {
        return plan == SubscriptionPlan.Pro ? INDIVIDUAL_PRO : INDIVIDUAL_FREE;
    }

    public static PlanLimit forOrgPlan(SubscriptionPlan plan) {
        if (plan == SubscriptionPlan.Enterprise) {
            throw new IllegalArgumentException("Enterprise limits are stored on OrgSubscription.");
        }
        return plan == SubscriptionPlan.Team ? ORG_TEAM : ORG_FREE;
    }
}
